### PR TITLE
Implement constructor type guard

### LINKED ISSUE #23274: Feature Request / Proposal: constructor type guard
In most cases, constructor comparison is faster than instanceof and typeof, thus using it as type guard can improve performance. The code below demonstrates the desired behavior:

```ts
const el:HTMLElement = document.createElement('canvas');

if (el.constructor === HTMLCanvasElement) {
    //inside of this scope, el is parsed as an instance of HTMLCanvasElement
    el.getContext('2d');
}
```

### PR DESCRIPTION
**Pull request by austincummings**
Fixes #23274
**tl;dr:** Implements constructor check type guards in the form of `x.constructor === T`.

---

This PR adds a new constructor equality check type guard. It specifically works on a set of binary expressions in the following syntactic forms:

- `x.constructor === T`
- `x.constructor == T`
- `x["constructor"] === T`
- `x["constructor"] == T`

Narrowing is done via the added `narrowTypeByConstructor` function. This type  guard only operates on `==` and `===` so if it finds the expression is using a  not-equals operator then it returns the type without any narrowing done.

After that it uses `isMatchingReference` to presumably check that we are still referring to the `x` part of the `x.constructor === T` expression. I'm not certain on how this works exactly, but the use case is to ensure only `x` has the narrowed type in the if body as opposed to `x.y` having the narrowed type.

Once that check is complete it moves on to getting the type referred to by `T`. This type eventually turns into our `candidate` type if it is not a direct reference to the `globalObjectType` or `globalFunctionType`. This does introduce a shortcoming of the narrowing as it prevents narrowing using a `T` of `Object`  or `Function`.

If the type that is being narrowed is `any`, then it can assume the `candidate` type is a subtype of `any` and so we just narrow directly to `candidate`. I  think this is a safe assumption but maybe not? Also there is no need to check for `unknown` as you would receive an error attempting to access `x.constructor` if `x`'s type is `unknown`.

Finally it returns the result of the `filterType` function which uses a function `isConstructedBy` to do the filtering. If none of the types match we end up with `never` as the flow type.

#### `isConstructedBy`
This function checks if either the `source` or `target` type are Object types with the `Class` `ObjectFlag`. If this is the case then we return if the symbols of `source` and `target` type are `===`. This is needed because you may have two classes with the same structure and TypeScript will see them as the same type. There was no way I could find that would check this. I think this also introduces the side effect of having pseudo nominal typing in the language.

For all other cases it just checks that the `source` type is a subtype of `target` type.

### Caveats and questions
- Does not narrow when using a `T` of `Object` or `Function`.
- Is comparing the symbols of `source` and `target` types a good way to check that they are referring to the exact same class?

### CODE REVIEW

**Comment by RyanCavanaugh** [comment]
@weswigham this LGTM - please merge if you agree

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -17336,6 +17336,12 @@ namespace ts {
                         if (right.kind === SyntaxKind.TypeOfExpression && isStringLiteralLike(left)) {
                             return narrowTypeByTypeof(type, <TypeOfExpression>right, operator, left, assumeTrue);
                         }
+                        if (isConstructorAccessExpression(left) && right.kind === SyntaxKind.Identifier) {
+                            return narrowTypeByConstructor(type, left, operator, right, assumeTrue);
+                        }
+                        if (isConstructorAccessExpression(right) && left.kind === SyntaxKind.Identifier) {
```
Comment (weswigham) [inline src/compiler/checker.ts]: I don't see why in `<expr>.constructor === <expr>` we'd limit what that the syntactic form of `<expr>` was. We're just calling `getTypeOfExpression` anyway.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -17594,6 +17600,59 @@ namespace ts {
                 return getTypeWithFacts(mapType(type, narrowTypeForTypeofSwitch(impliedType)), switchFacts);
             }
 
+            function narrowTypeByConstructor(type: Type, constructorAccessExpr: AccessExpression, operator: SyntaxKind, identifier: Expression, assumeTrue: boolean): Type {
+                // Do not narrow when checking inequality.
+                if (!assumeTrue || operator !== SyntaxKind.EqualsEqualsToken && operator !== SyntaxKind.EqualsEqualsEqualsToken) {
```
Comment (weswigham) [inline src/compiler/checker.ts]: This condition is incorrect. If `!assumeTrue`, we should perform the narrowing only for `!==` and `!=`, while if `assumeTrue`, we should perform narrowing for `===` and `==`. So:

```suggestion
                if (assumeTrue ? (operator !== SyntaxKind.EqualsEqualsToken && operator !== SyntaxKind.EqualsEqualsEqualsToken) : (operator !== SyntaxKind.ExclamationEqualsToken && operator !== SyntaxKind.ExclamationEqualsEqualsToken)) {
```

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -17594,6 +17600,59 @@ namespace ts {
...
+                }
+
+                // In the case of `x.y`, a `x.constructor === T` type guard resets the narrowed type of `y` to its declared type.
+                if (!isMatchingReference(reference, constructorAccessExpr.expression)) {
+                    return declaredType;
+                }
+
+                // Get the type of the constructor identifier expression, if it is not a function then do not narrow.
+                const identifierType = getTypeOfExpression(identifier);
+                if (!isTypeSubtypeOf(identifierType, globalFunctionType)) {
```
Comment (weswigham) [inline src/compiler/checker.ts]:
```suggestion
                if (!isFunctionType(identifierType) && !isConstructorType(identifierType)) {
```
this is much different, but aligns much more with what you'd like to check (and should be faster, to boot).

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -3990,6 +3990,13 @@ namespace ts {
         return isPropertyAccessExpression(node) && isEntityNameExpression(node.expression);
     }
 
+    export function isConstructorAccessExpression(expr: Expression): expr is AccessExpression {
+        return (
+            isPropertyAccessEntityNameExpression(expr) && idText(expr.name) === "constructor" ||
```
Comment (weswigham) [inline src/compiler/utilities.ts]: I'd loosen this - there's no need for whatever's left of the `constructor` to be so strict, so long as the references match. (Plus, as written this prevents something like `this[mySymbol].constructor === Cls` from working)
```suggestion
            isPropertyAccessExpression(expr) && idText(expr.name) === "constructor" ||
```
especially since the `ElementAcessExpression` case below is already more general and allows arbitrary LHS expressions.

File: tests/baselines/reference/typeGuardConstructorClassAndNumber.errors.txt [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        var1.property1; // string
+    }
+    
+    if (var1.constructor != C1) {
+        var1; // C1 | number
+        var1.property1; // error
+             ~~~~~~~~~
+!!! error TS2339: Property 'property1' does not exist on type 'number | C1'.
+!!! error TS2339:   Property 'property1' does not exist on type 'number'.
+    }
```
Comment (weswigham) [inline tests/baselines/reference/typeGuardConstructorClassAndNumber.errors.txt]: After you do my first comment's change, adding `else` branches to all these is pertinent.

File: src/compiler/checker.ts (hunk 1 above)
Comment (austincummings) [reply src/compiler/checker.ts]: Good point. I'll change this.

File: src/compiler/checker.ts (hunk 2 above)
Comment (austincummings) [reply src/compiler/checker.ts]: @weswigham I think you're right and I'll make this change as well. 

After looking at this a bit more though, could we be performing narrowing in the false cases too?

For example, given:
```ts
class C1 {}

let var1: C1 | number;
if (var1.constructor == C1) {
    var1; // C1
}
else {
    var1; // number | C1
}

if (var1.constructor != C1) {
    var1; // number | C1
}
else {
    var1; // C1
}
```
In the first `else` body, and the second `if` body, can we remove `C1` from the union since (I think) we know the type can't ever be `C1`?

File: src/compiler/checker.ts (hunk 3 above)
Comment (austincummings) [reply src/compiler/checker.ts]: Will make this change. Thanks for the heads up on that.

File: src/compiler/utilities.ts (hunk 4 above)
Comment (austincummings) [reply src/compiler/utilities.ts]: Ah good point, thanks. I'll change this.

File: tests/baselines/reference/typeGuardConstructorClassAndNumber.errors.txt (hunk 5 above)
Comment (austincummings) [reply tests/baselines/reference/typeGuardConstructorClassAndNumber.errors.txt]: Will do 👍

File: src/compiler/checker.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/checker.ts]: No - given another interface structurally identical to C1, but with a different `constructor` symbol, we'd fall into the false branch, but still "have a C1".

**Comment by austincummings** [comment]
Addressed @weswigham's comments.

**Comment by weswigham** [comment]
Looks like some tests are failing - you may need to run `--skipPercent=0` to prevent the test from being skipped in a test run locally.

**Comment by austincummings** [comment]
Ah ok, I forgot to rebase onto master. Looks like some symbol positions changed.

**Comment by weswigham** [comment]
So, showerthought here, but.... What do we do when
```ts
const x = { constructor: String };
if (x.constructor === String) {
    x.trim(); // runtime error
}
```
Or any other instance where the constructor property isn't inherited from `Object` or `Function`? I think we need to validate that the constructor property accessed comes from one of the built-in apparent types for this narrowing to be safe to do.

**Comment by austincummings** [comment]
@weswigham, just tried it out with my local build and it looks like that 
specific case won't be an issue. `x` will have a type of `never` inside the 
if so you will get a compile time error. This is because we aren't 
comparing against the type of `x`'s constructor, but rather against the 
type of `x`, and the prototype type of the identifier right of the 
triple-equals directly.

I haven't thought of a case yet where this will break but I'll definitely 
keep thinking about it.

**Comment by matthew-dean** [comment]
I came here to request this feature, as I'm doing `.constructor` types extensively in my project, so this is great! Can't wait for this to make it in!

**Comment by KingDarBoja** [comment]
> I came here to request this feature, as I'm doing `.constructor` type checks extensively in my project (but TypeScript still needs type assertions), so this is great! Can't wait for this to make it in!

Same here, this is really an awesome feature and glad the author and the Typescript team going to support it :cake:

**Comment by levidrain** [comment]
Is this slated to be released soon?

**Comment by hayd** [comment]
Bump!

**Comment by darsain** [comment]
I'd like to suggest adding an alternative `x.constructor.name === "T"` typeguard.

Why / where am I coming from:

I worked once in a weird environment (I think an nw.js app) where my element objects were defined in a different context than globally available constructors, so stuff like

```js
actualInputElement instanceof HTMLInputElement
```

would always return `false`. Since than I consider `instanceof` or comparing raw constructor objects on stuff like HTML elements unsafe and do `constructor.name` checks there instead, as that was the only alternative.

Is it too unsound to even be considered? :) Because it's working perfectly on html elements and I don't have to be afraid of weird environments with colliding contexts.

**Comment by HolgerJeromin** [comment]
@darsain 
> I worked once in a weird environment (I think an nw.js app) where my element objects were defined in a different context than globally available constructors, so stuff like
> `actualInputElement instanceof HTMLInputElement`
> would always return false. 

I had a similar situation inside an HTMLObjectElement. I think 
`actualInputElement instanceof actualInputElement.ownerDocument.defaultView.HTMLInputElement`
is a saver runtime check, but no TS check either.
